### ISSUE TITLE
Parallel processing during the backward pass

### ISSUE THREAD

**Issue by wxmerkt** [loco-3d/crocoddyl#113]
In GitLab by @proyan on Feb 14, 2019, 13:56

This is not a task (yet). I am just reporting on what I did.

I tried the multiprocessing library of python for doing parallel processing of the backward pass. Then I tried the multi-threading library. Neither of these worked, and here is why:
-----------------------------------------
Multiprocessing library uses os.fork() to create new child processes. os.fork() copies the entire state of the interpreter, and copies it into the children. When the processing is done, the children vanish, and any data that is not returned also vanishes. 

Unfortunately, since we are not doing any memory allocation, and doing all changes within the data object of various classes, any new computation done within the data elements of the children processes will vanish with the processes. Only the returned objects will be copied back to the parent process.

We have three options: a) implement our own fork() which does not copy the environment, b) return the new data elements and copy them into the parent process. c) use a shared memory to share the data between the multiple processes

I didn't do (a) because it demanded a lot of time. For (b) is not really efficient, and (b) and (c) require that the data elements be picklable, which is not the case for the pinocchio objects.
----------------------------------------
Multithreading library does not copy the memory, and runs threads instead of processes, which is good. However, there is the Global Interpreter Lock(GIL) of python, which forces that only one thread can manipulate the interpretor memory at any given time. Which is stupid. It means that you don't really run multithreading, you run multiple threads in sequence and with an overhead.

This is indeed what I noticed. 1 iteration of the biped example took me 3s. 1 iteration with 4 threads took 3.2s, and 1 iteration with 8 threads took 3.5s.
--------------------------------------------------

In conclusion, if anyone wants to try out the python multiprocessing, keep this issue in mind and work from where I left off. 

Otherwise, I think we have to wait for the cpp implementation before we can think about parallel processing.

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Feb 14, 2019, 14:02

Concerning the Pickling of Pinocchio Data, it would be possible to do it.
I have implemented Pickling for Spatial objects. It is then a matter of continuing this job.

By the way, Pickling Data can be shorten as the creation of a new Data object.

**Comment by wxmerkt** [comment]
In GitLab by @nmansard on Feb 14, 2019, 14:03

Clearly, we should not go with multiprocess, but with multi thread. If not possible in Python ... too bad. 
Could you copy the (failed) test with multithread somewhere for archive?

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Feb 15, 2019, 15:16

Great to report this to us. I tend to think that we should keep this task for c++ implementation. Python for my is more about analysing the scientific computation, instead in c++ we should carefully address the software efficiency (and this kind of things).

**Comment by wxmerkt** [comment]
In GitLab by @proyan on Feb 18, 2019, 13:29

assigned to @proyan

**Comment by wxmerkt** [comment]
In GitLab by @nmansard on Feb 19, 2019, 20:36

@proyan is pushing for multiprocess as multithread is not feasible in Python. https://github.com/stack-of-tasks/pinocchio/pull/682 is one step in this direction. WIP.

**Comment by wxmerkt** [comment]
In GitLab by @nmansard on Feb 28, 2019, 11:17

unassigned @proyan

**Comment by wxmerkt** [comment]
In GitLab by @nmansard on Mar 4, 2019, 10:56

@proyan tried and it is more complex that expected. As it is not relevant for the C++ version (that will not use multiprocess anyway), we will stop exploring this direction.

@proyan, please record the result of your research here.

**Comment by wxmerkt** [comment]
In GitLab by @nmansard on Mar 4, 2019, 10:56

closed

**Comment by wxmerkt** [comment]
In GitLab by @proyan on Mar 4, 2019, 11:48

The report on what I did:

Since multitasking cannot work on python for a CPU-intensive task, I decided to try multiprocessing for the issue. multiprocessing on python requires pickling of the global dict, which requires pickling of pinocchio quantities as well (model and data). I worked on that two weeks ago, and exposed pinocchio::JointData for this purpose.

However, on Friday, I looked on the pickling again after discussion with Justin about the pickling performance.
1) Looking at the pickling performance when compared to the expected benefits, I decided that pickling might not be the best way forward.
2) Looking at the multiprocessing module, I found that it might be possible to use shared memory without pickling, where we can store the models and datas directly inside the multiprocessing heap.

While I did not have numerical proof of the pickling performance, (2) offered much better performance, so I investigated it further.

Working on (2) would require
* A constructor in pinocchio::Model and pinocchio::Data which allows us to create an object at a given memory location
* Updating the forking of multiprocessing to avoid pickling of the shared memory
* implementing this in crocoddyl.

This is too intensive work, much of which is concentrated in the multiprocessing modul. Also, this is prone to bugs. I investigated the issue for one day, but I don't think I should spend more time on it.

Looking at the cost-benefit of including multiprocessing in python, I'm not pursuing this anymore. I'll revisit this topic after we have the c++ implementation.
